server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java

```diff
@@ -0,0 +1,648 @@
+            assertThat(shardAllocationDecision.isDecisionTaken(), equalTo(true));
+            assertThat(shardAllocationDecision.getAllocateDecision().getAllocationDecision(),
+                equalTo(AllocationDecision.NO_VALID_SHARD_COPY));
+        });
+
+        internalCluster().restartNode(node, new InternalTestCluster.RestartCallback() {
+            @Override
+            public Settings onNodeStopped(String nodeName) throws Exception {
+                t.addTextInput("y");
+
+                // Try running it before the shard is corrupted, it should flip out because there is no corruption file marker
+                command.execute(t, options, environment);
+
+                return super.onNodeStopped(nodeName);
+            }
+        });
+
+        waitNoPendingTasksOnAll();
+
+        String nodeId = null;
+        final ClusterState state = client().admin().cluster().prepareState().get().getState();
+        final DiscoveryNodes nodes = state.nodes();
+        for (ObjectObjectCursor<String, DiscoveryNode> cursor : nodes.getNodes()) {
+            final String name = cursor.value.getName();
+            if (name.equals(node)) {
+                nodeId = cursor.key;
+                break;
+            }
+        }
+        assertThat(nodeId, notNullValue());
+
+        logger.info("--> output:\n{}", t.getOutput());
+
+        assertThat(t.getOutput(), containsString("allocate_stale_primary"));
+        assertThat(t.getOutput(), containsString("\"node\" : \"" + nodeId + "\""));
+
+        // there is only _stale_ primary (due to new allocation id)
+        assertBusy(() -> {
+            final ClusterAllocationExplanation explanation =
+                client().admin().cluster().prepareAllocationExplain()
+                    .setIndex(indexName).setShard(0).setPrimary(true)
+                    .get().getExplanation();
+
+            final ShardAllocationDecision shardAllocationDecision = explanation.getShardAllocationDecision();
+            assertThat(shardAllocationDecision.isDecisionTaken(), equalTo(true));
+            assertThat(shardAllocationDecision.getAllocateDecision().getAllocationDecision(),
+                equalTo(AllocationDecision.NO_VALID_SHARD_COPY));
+        });
+
+        client().admin().cluster().prepareReroute()
+            .add(new AllocateStalePrimaryAllocationCommand(indexName, 0, nodeId, true))
+            .get();
+
+        assertBusy(() -> {
+            final ClusterAllocationExplanation explanation =
+                client().admin().cluster().prepareAllocationExplain()
+                    .setIndex(indexName).setShard(0).setPrimary(true)
+                    .get().getExplanation();
+
+            assertThat(explanation.getCurrentNode(), notNullValue());
+            assertThat(explanation.getShardState(), equalTo(ShardRoutingState.STARTED));
+        });
+
+        final Pattern pattern = Pattern.compile("Corrupted Lucene index segments found -\\s+(?<docs>\\d+) documents will be lost.");
```

Comment:
I ran it so many times (more than several thousands) - and there are only two options 
 - or `CheckIndex` provides correct number of docs in a corrupted segments - or segments are missed and it can not figure out on how to glue the rest.